src/basicfuns.jl

```diff
@@ -63,7 +63,7 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref). 
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
+log1pexp(x::Real) = x ≤ -37. ? exp(x) : x ≤ 18. ? log1p(exp(x)) : x ≤ 33.3 ? x + exp(-x) : float(x)
```

Comment:
I wonder if it would be better to preserve the `oftype`, since in practice not all subtypes of `Real` implement `float` properly (in particular, some AD packages).
```suggestion
log1pexp(x::Real) = x ≤ -37. ? exp(x) : x ≤ 18. ? log1p(exp(x)) : x ≤ 33.3 ? x + exp(-x) : oftype(exp(-x), x)
```

Reply:
Actually, probably better to do something along the lines of:
```suggestion
log1pexp(x::Real) = log1pexp(float(x))
log1pexp(x::Float64) = x ≤ -37.0 ? exp(x) : 
                       x ≤ 18.0 ? log1p(exp(x)) :
                       x ≤ 33.3 ? x + exp(-x) :
                       x
```

Reply:
> Actually, probably better to do something along the lines of:

But would that work with the AD packages @tpapp is referring to? It still depends on `float`.
As for the `oftype(exp(-x), x)` call, it has the downside that it performs an extra `exp` call which can be slow.

Reply:
I thought the compiler would elide `exp`, and just figure out the type, but you are right, it remains in there.

Perhaps `float` is OK, and the packages not supporting it should just be fixed.

Reply:
So we agree that `float` is OK? Any issues with this or it can be merged?